### DISCUSSION TITLE
How to use a React hook instead of a Component in Reflex

### DISCUSSION THREAD

**Discussion by scottpersinger** [reflex-dev/reflex#2621] [General]
> **Answered**:
> Looks like support was added for React Hooks, but I guess the docs are quiet because this answer was not easily Googleable: https://github.com/reflex-dev/reflex/pull/810.
>
> I am filing this discussion in case it comes up for anyone else.

It seems that the [@react-oauth/google](https://www.npmjs.com/package/@react-oauth/google) supports the Google [Authorization code flow](https://www.npmjs.com/package/@react-oauth/google#authorization-code-flow) but only via a React hook, it's not built into the component. So the example in JS says:

```js
import { useGoogleLogin } from '@react-oauth/google';

const login = useGoogleLogin({
  onSuccess: codeResponse => console.log(codeResponse),
  flow: 'auth-code',
});

<MyCustomButton onClick={() => login()}>Sign in with Google 🚀</MyCustomButton>;
```

I am wondering if it's possible to use this style in Reflex, as the [examples](https://github.com/reflex-dev/reflex-examples/blob/main/google_auth/google_auth/react_oauth_google.py) all point to wrapping a full Component:

```python
class GoogleLogin(rx.Component):
    library = "@react-oauth/google"
    tag = "GoogleLogin"

    def get_event_triggers(self):
        return {"on_success": lambda data: [data]}
```

**Comment by masenf** [comment]
Is that example code supposed to be inside a component?

At any rate, override `_get_hooks` to return JS code that will be inserted inside the component.

```python
class GoogleLoginButton(rx.chakra.Button):
    @classmethod
    def create(cls, **props):
        return super().create("Sign in with Google 🚀", **props)

    def _get_imports(self) -> rx.utils.imports.ImportDict:
        return rx.utils.imports.merge_imports(
            super()._get_imports(),
            {
                "@react-oauth/google": {rx.utils.imports.ImportVar(tag="useGoogleLogin")}
            },
        )

    def _get_hooks(self) -> str | None:
        return """
const login = useGoogleLogin({
  onSuccess: codeResponse => %s(codeResponse),
  flow: 'auth-code',
});""" % self.event_triggers["on_success"]

    def _render(self):
        """Remove the onSuccess prop so that it doesn't try to render to the DOM."""
        tag = super()._render()
        tag.remove_props("onSuccess")
        tag.add_props(on_click="() => login()")
        return tag

    def get_event_triggers(self):
        return {"on_success": lambda data: [data]}
```

The problem is that this doesn't work inside an `rx.cond` due to #2606, which needs to be fixed.
